akka-cluster-sharding/src/main/scala/akka/cluster/sharding/external/ExternalShardAllocationStrategy.scala

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (C) 2019-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.external
+
+import akka.actor.Actor
+import akka.actor.ActorLogging
+import akka.actor.ActorRef
+import akka.actor.ActorRefScope
+import akka.actor.Address
+import akka.actor.AddressFromURIString
+import akka.actor.ClassicActorSystemProvider
+import akka.actor.ExtendedActorSystem
+import akka.actor.NoSerializationVerificationNeeded
+import akka.actor.Props
+import akka.actor.Stash
+import akka.cluster.Cluster
+import akka.cluster.ddata.DistributedData
+import akka.cluster.ddata.LWWMapKey
+import akka.cluster.ddata.Replicator.Changed
+import akka.cluster.ddata.Replicator.Subscribe
+import akka.cluster.sharding.ShardCoordinator
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.event.Logging
+import akka.pattern.AskTimeoutException
+import akka.util.Timeout
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+object ExternalShardAllocationStrategy {
+
+  type ShardRegion = ActorRef
+
+  // local only messages
+  private[akka] final case class GetShardLocation(shard: ShardId)
+  private[akka] final case object GetShardLocations
+  private[akka] final case class GetShardLocationsResponse(desiredAllocations: Map[ShardId, Address])
+  private[akka] final case class GetShardLocationResponse(address: Option[Address])
+
+  // only returned locally, serialized as a string
+  final case class ShardLocation(address: Address) extends NoSerializationVerificationNeeded
+
+  private object DDataStateActor {
+    def props(typeName: String) = Props(new DDataStateActor(typeName))
+  }
+
+  // uses a string primitive types are optimized in ddata to not serialize every entity
+  // separately
+  private[akka] def ddataKey(typeName: String): LWWMapKey[ShardId, String] = {
+    LWWMapKey[ShardId, String](s"external-sharding-$typeName")
+  }
+
+  private class DDataStateActor(typeName: String) extends Actor with ActorLogging with Stash {
+
+    private val replicator = DistributedData(context.system).replicator
+
+    private val Key = ddataKey(typeName)
```

Comment:
We've gone back and forth with this (see history), it did shard across a configurable number of keys, then a hard coded number, and finally after some testing showing it can handles 10s of thousands of shards with a single key I put it back to a single key.
We can change this later without it being a breaking change so we're not fixed to a singl ekey.

Reply:
Perhaps mention those 10s of thousands of shards in the doc section then to set the expectations?

Reply:
:+1: